// sucrase/dist/register.mjs
// @ts-ignore: no types available.
import * as pirates from "pirates";
import { transform} from "./index";

export function addHook(extension, options) {
  pirates.addHook(
    (code, filePath) => transform(code, {...options, filePath}).code,
    {exts: [extension]},
  );
}

export function registerJS() {
  addHook(".js", {transforms: ["imports", "flow", "jsx"]});
}

export function registerJSX() {
  addHook(".jsx", {transforms: ["imports", "flow", "jsx"]});
}

export function registerTS() {
  addHook(".ts", {transforms: ["imports", "typescript"]});
}

export function registerTSX() {
  addHook(".tsx", {transforms: ["imports", "typescript", "jsx"]});
}

export function registerTSLegacyModuleInterop() {
  addHook(".ts", {
    transforms: ["imports", "typescript"],
    enableLegacyTypeScriptModuleInterop: true,
  });
}

export function registerTSXLegacyModuleInterop() {
  addHook(".tsx", {
    transforms: ["imports", "typescript", "jsx"],
    enableLegacyTypeScriptModuleInterop: true,
  });
}

export function registerAll() {
  registerJS();
  registerJSX();
  registerTS();
  registerTSX();
}
